rust/src/http2/detect.rs

```diff
@@ -1000,12 +1000,31 @@ fn http2_caseinsensitive_cmp(s1: &[u8], s2: &str) -> bool {
     return false;
 }
 
+// This macro builds a rust slice from a C pointer
+// It handles the case where the C pointer is NULL by creating an empty slice
+// This is not the idiomatic rust behavior
+// But it is fit for this specific use case, hence the macro is not public
+// Every HTTP2 header has a name and a value, even if an empty one
+// It is not currently guaranteed with HTTP1 headers
+// During protocol upgrade, we translate a HTTP1 request with headers into a HTTP2 one
+// So, we translate an absent name/value into an empty one
+macro_rules! build_slice_null {
+    ($buf:ident, $len:expr) => {
+        if $buf.is_null() {
+            &[]
+        } else {
+            // unsafe
+            std::slice::from_raw_parts($buf, $len)
+        }
+    };
+}
```

Comment:
> Can we convert this to a utility function?

I managed to hack something... 

>  I don't think there is any reason for it to be a macro.

Because the dummy transcription into a function does not compile because of lifetime and such...

> It could then be used in applayer.rs as well which implements the same behavior.

It is not exactly the same behavior (check for length)

Reply:
Actually this commit does not seem needed as libhtp does `h->name = bstr_dup_c("");` if there is no name.

Better proof : `#define bstr_ptr(X) ( ((*(X)).realptr == NULL) ? ((unsigned char *)(X) + sizeof(bstr)) : (unsigned char *)(*(X)).realptr )`